### PR TITLE
[OptimizeInstructions] Canonicalizations for relational ops near signed & unsigned min/max

### PR DESCRIPTION
**Pull request by MaxGraey**
It's continue of #4272
```rust
(signed)x  < s_min + 1   ==>   x == s_min
(signed)x >= s_min + 1   ==>   x != s_min
(signed)x  > s_max - 1   ==>   x == s_max
(signed)x <= s_max - 1   ==>   x != s_max

(unsigned)x <= u_max - 1   ==>   x != u_max
(unsigned)x  > u_max - 1   ==>   x == u_max
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
@kripken @tlively Could you take a look to this please?

**Comment by MaxGraey** [comment]
@tlively Could you take a little look at this old PR please?

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -1985,6 +1985,71 @@ struct OptimizeInstructions
         c->value = Literal::makeZero(c->type);
         return;
       }
+      // Prefer compare to signed min (s_min) instead of s_min + 1.
+      // (signed)x < s_min + 1   ==>   x == s_min
+      if (binary->op == LtSInt32 && c->value.geti32() == -2147483647L) {
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:1990]: How about using `INT32_MIN + 1` here and similar constants instead of hardcoded numbers elsewhere?

**Review by tlively — COMMENTED** [review]
LGTM with that one comment.

File: src/passes/OptimizeInstructions.cpp (hunk 1 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:1990]: Good idea!

**Review by tlively — APPROVED** [review]

**Comment by MaxGraey** [comment]
@tlively Thanks!

File: test/lit/passes/optimize-instructions.wast [hunk 2]
```diff
@@ -1644,6 +1644,136 @@
...
+    ;; (signed)x <= s_max - 1   ==>   x != s_max
+    (drop (i32.le_s
+      (local.get $x)
+      (i32.const 2147483646)
+    ))
+    (drop (i64.le_s
+      (local.get $y)
+      (i64.const 9223372036854775806)
+    ))
+    ;; (unsigned)x <= u_max - 1   ==>   x == u_max
```
Comment (kripken) [inline test/lit/passes/optimize-instructions.wast:1758]: The text here is wrong, `==` should be `!=` (but the optimizer output is right).

Same in the comment below, `!=` should be `==`.

Comment (MaxGraey) [reply test/lit/passes/optimize-instructions.wast:1758]: Thanks! Good catch. Do you want a separate PR which fix this comment or just append this minor fix with some other existing PR?

Comment (kripken) [reply test/lit/passes/optimize-instructions.wast:1758]: A separate PR might be simplest. I am not up to date on the other PRs and am not sure when they will land. A separate PR can be landed immediately.

Comment (MaxGraey) [reply test/lit/passes/optimize-instructions.wast:1758]: https://github.com/WebAssembly/binaryen/pull/4992

**Pull request merged**